File: vigenere.py
import cs50
import sys

def main():
    
    # Error if more or less than 1 command line argument
    if len(sys.argv) != 2:
        print("Error. Usage: python vigenere.py key")
        exit(1)
        
    m = len(sys.argv[1])
    k = sys.argv[1]
    
    # Converts key from ASCII to alphabetical indices    
    for d in k:
        if not str.isalpha(d):
            print("Error. Use a key with only alphabets.")
            exit(1)
    
    # Get plaintext from user
    print("plaintext: ", end="")
    p = cs50.get_string()

    
    if p != None:
        print("ciphertext: ", end="")
        
        j = 0
        
        for c in p:
            if j > m - 1:
                j = 0
            if str.isalpha(c):
                if str.isupper(c):
                    if str.isupper(k[j]):
                        c = (ord(c) - ord('A') + ord(k[j]) - ord('A')) % 26
                        print(chr(c + ord('A')), end="")
                        j += 1
                    else:
                        c = (ord(c) - ord('A') + ord(k[j]) - ord('a')) % 26
                        print(chr(c + ord('A')), end="")
                        j += 1
                else:
                    if str.isupper(k[j]):
                        c = (ord(c) - ord('a') + ord(k[j]) - ord('A')) % 26
                        print(chr(c + ord('a')), end="")
                        j += 1
                    else:
                        c = (ord(c) - ord('a') + ord(k[j]) - ord('a')) % 26
                        print(chr(c + ord('a')), end="")
                        j += 1
            else:
                print(c, end="")
        
        print()
    
if __name__ == "__main__":
    main()

File: vigenere.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <stdlib.h>
#define MODC 26
#define ASCUP 65
#define ASCLO 97

int main(int argc, string argv[])
{
    // error if more or less than one command line argument 
    if (argc != 2)
    {
        printf("Error. Usage: ./vigenere key\n");
        return 1;
    }
    
    // making sure argument string is valid
    if (argv[1] != NULL)
    {
        int m = strlen(argv[1]);
    
        // converts key from ASCII to alphabetical indices. Error if key has characters other than alphabets.
        for (int j = 0; j < m; j++)
        {
            if (isalpha(argv[1][j]))
            {
                if (isupper(argv[1][j]))
                {
                    argv[1][j] = argv[1][j] - ASCUP;
                }
                else if (islower(argv[1][j]))
                {
                    argv[1][j] = argv[1][j] - ASCLO;
                }
            }
            else
            {
                printf("Error. Use a key with only alphabets.\n");
                return 1;
            }
        }
        
        // get plaintext from user
        printf("plaintext: ");
        string p = get_string();
        
        // making sure entered string is valid
        if (p != NULL)
        {
            printf("ciphertext: ");
            
            // going through the string one character at a time. 
            // In this loop, j is only incremented if the ith character in p is an alphabet.
            
            for (int i = 0, n = strlen(p), j = 0; i < n; i++)
            {
                // if character is an alphabet
                if (isalpha(p[i]))
                {
                    // if alphabet is uppercase. 
                    if (isupper(p[i]))
                    {
                        p[i] = (p[i] - ASCUP + argv[1][(j % m)]) % MODC;
                        printf("%c", p[i] + ASCUP);
                        j++;
                    }
                    // if alphabet is lowercase
                    else
                    {
                        p[i] = (p[i] - ASCLO + argv[1][(j % m)]) % MODC;
                        printf("%c", p[i] + ASCLO);
                        j++;
                    }
                    
                }
                // if not an alphabet, print as it is
                else
                    printf("%c", p[i]);
            }
            // print a new line at the end
            printf("\n");
        }
    }
}